### PR TITLE
Add create job command to create job from cronjob

### PR DESCRIPTION
**Pull request by mpon**
### What this PR does

Added `kubectl fuzzy create-job-from` command.
I want to create job from cronjob by fuzzy finder which is equivalent to following kubectl command.

```bash
kubectl create job my-job --from cronjob/my-cronjob
```

to

```bash
// generated job name from cronjob
kubectl fuzzy create-job-from cronjob
// create job by specified job name
kubectl fuzzy create-job-from cronjob my-specified-job-name
```

### Demo

cj

#### 1. command name

I'm not sure this command name is good.
I made this `kubectl fuzzy create-job-from-cronjob` first, but if create job command will be supported other than cronjob, it will have to add another command. So I passed resource as argument.

#### 2. default preview mode

When I created a job from a cronjob, I thought it was better to see only the information that get cronjob had, rather than looking at the cronjob yaml.

### *Edit after discussion*

* change command name: `kubectl fuzzy create job --from=cronjob -P`
* use existing preview mode option
* use `fuzzyfinder.Infos` to find cronJob, and cast it to CronJob

fuzzy_cj

### CODE REVIEW

File: pkg/cmd/create_job_from.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+	cronJobClient batchv1beta1client.CronJobsGetter
+	jobClient     batchv1client.JobsGetter
+	namespace     string
+}
+
+// NewCreateJobFromOptions provides an instance of CreateJobFromOptions with default values.
+func NewCreateJobFromOptions(streams genericclioptions.IOStreams) *CreateJobFromOptions {
+	return &CreateJobFromOptions{
+		configFlags: genericclioptions.NewConfigFlags(true),
+		printFlags:  genericclioptions.NewPrintFlags("created").WithTypeSetter(scheme.Scheme),
```
Comment (mpon) [inline pkg/cmd/create_job_from.go:85]: reference from https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/create/create_job.go#L79-L84

File: pkg/cmd/create_job_from.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+	o.namespace = namespace
+
+	printer, err := o.printFlags.ToPrinter()
+	if err != nil {
+		return err
+	}
+
+	o.printObj = func(obj runtime.Object) error {
+		return printer.PrintObj(obj, o.Out)
+	}
```
Comment (mpon) [inline pkg/cmd/create_job_from.go:131]: reference from https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/create/create_job.go#L155-L161

File: pkg/cmd/create_job_from.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+	res, err := o.jobClient.Jobs(cronJob.Namespace).Create(context.Background(), job, createOptions)
+
+	if err != nil {
+		return fmt.Errorf("failed to create job: %v", err)
+	}
+
+	return o.printObj(res)
+}
+
+func (o *CreateJobFromOptions) createJobFromCronJob(cronJob *batchv1beta1.CronJob, name *string) *batchv1.Job {
```
Comment (mpon) [inline pkg/cmd/create_job_from.go:169]: reference from https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/create/create_job.go#L263

File: pkg/fuzzyfinder/fuzzyfinder.go [hunk 4]
```diff
@@ -179,3 +183,53 @@ func convert(jsonObj []byte, printer printers.ResourcePrinter) (string, error) {
...
+				translateTimestampSince(cj.CreationTimestamp),
+			)
+		}
+		return ""
+	})
+}
+
+// translateTimestampSince returns the elapsed time since timestamp in
+// human-readable approximation.
+func translateTimestampSince(timestamp metav1.Time) string {
```
Comment (mpon) [inline pkg/fuzzyfinder/fuzzyfinder.go:229]: reference from https://github.com/kubernetes/kubernetes/blob/master/pkg/printers/internalversion/printers.go#L670

**Comment by d-kuro** [comment]
@mpon 
Thanks for your contribution! Sorry for the delay in replying.

I had a hard time designing the subcommands, but I think better to kubectl fuzzy should be able to use the same interface as kubectl interface.

For example, I think it would look something like this:
`kubectl fuzzy create job foo --from=cronjob`

The `--from` the option will be required in this case, which may compromise usability, but it will be the same interface as kubectl, so you won't have to learn a new command.

Also, about the preview option.
I want to be consistent in this plugin so the default is no preview, and show YAML /JSON in `-P` or `—preview` option.

I think the preview you proposed is useful, but we will have to discuss what to show on objects other than Kubernetes' CronJob to keep it consistent.

**Comment by mpon** [comment]
Thank you for your consideration.
I think it's good to follow the kubectl interface and also the consistency of options in this plugin.
I'll try to modify subcommands and preview options.

File: pkg/cmd/create.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+func NewCmdCreate(streams genericclioptions.IOStreams) *cobra.Command {
+	o := NewCreateOptions(streams)
+
+	cmd := &cobra.Command{
+		Use:           "create",
+		Short:         "Create a resource",
+		SilenceUsage:  true,
+		SilenceErrors: true,
+		RunE: func(c *cobra.Command, args []string) error {
+			return fmt.Errorf("must specify resource, only supported job")
```
Comment (d-kuro) [inline pkg/cmd/create.go:23]:
```suggestion
			return errors.New("must specify resource, only supported job")
```

In this case, I think better to `errors.New`.
(you need a fix imports)

File: pkg/cmd/create_job.go [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+// Complete sets all information required for get logs.
+func (o *CreateJobOptions) Complete(cmd *cobra.Command, args []string) error {
+	client, err := kubernetes.NewClient(o.configFlags)
+	if err != nil {
+		return fmt.Errorf("failed to new Kubernetes client: %w", err)
+	}
+
+	if o.from == "" {
+		return fmt.Errorf("--from option is required, only supported job from cronjob")
```
Comment (d-kuro) [inline pkg/cmd/create_job.go:118]: 👍

Comment (d-kuro) [reply pkg/cmd/create_job.go:118]:
```suggestion
		return errors.New("--from option is required, only supported job from cronjob")
```

File: pkg/cmd/create_job.go [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+	}
+
+	if o.from != "cronjob" {
+		return fmt.Errorf("must specify resource, only supported job")
```
Comment (d-kuro) [inline pkg/cmd/create_job.go:122]:
```suggestion
		return errors.New("must specify resource, only supported cronjob")
```

File: pkg/cmd/create_job.go [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+	cronJob, err := fuzzyfinder.CronJobs(cronJobs.Items, printer, o.RawPreview)
+	if err != nil {
+		return fmt.Errorf("failed to fuzzyfinder execute: %w", err)
+	}
+
+	job := o.createJobFromCronJob(&cronJob, &o.name)
+
+	createOptions := metav1.CreateOptions{}
+	res, err := o.jobClient.Jobs(cronJob.Namespace).Create(context.Background(), job, createOptions)
+
```
Comment (d-kuro) [inline pkg/cmd/create_job.go:182]:
```suggestion
```

File: pkg/cmd/create_job.go [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+	cronJob, err := fuzzyfinder.CronJobs(cronJobs.Items, printer, o.RawPreview)
+	if err != nil {
+		return fmt.Errorf("failed to fuzzyfinder execute: %w", err)
+	}
+
+	job := o.createJobFromCronJob(&cronJob, &o.name)
+
+	createOptions := metav1.CreateOptions{}
+	res, err := o.jobClient.Jobs(cronJob.Namespace).Create(context.Background(), job, createOptions)
```
Comment (d-kuro) [inline pkg/cmd/create_job.go:181]:
```suggestion
	res, err := o.jobClient.Jobs(cronJob.Namespace).Create(context.Background(), job, metav1.CreateOptions{})
```

I think you can sum it up in one line.

File: pkg/cmd/create_job.go [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+
+	if err != nil {
+		return fmt.Errorf("failed to create job: %v", err)
```
Comment (d-kuro) [inline pkg/cmd/create_job.go:184]:
```suggestion
		return fmt.Errorf("failed to create job: %w", err)
```

I think better for wrapping errors.

File: pkg/fuzzyfinder/fuzzyfinder.go [hunk 11]
```diff
@@ -206,3 +207,69 @@ func multipleGVKsRequested(infos []*resource.Info) bool {
 
 	return false
 }
+
+func rawCronJobPreviewWindow(cronJobs []batchv1beta1.CronJob, printer printers.ResourcePrinter) fuzzyfinder.Option {
+	gvk := schema.GroupVersionKind{Kind: "Pod", Version: "v1"}
```
Comment (d-kuro) [inline pkg/fuzzyfinder/fuzzyfinder.go:212]:
```suggestion
	gvk := schema.GroupVersionKind{Group: "batch", Version: "v1beta1", Kind: "CronJob"}
```

File: pkg/cmd/create.go (hunk 5 above)
Comment (mpon) [reply pkg/cmd/create.go:23]: Thanks! I'll fix it.

File: pkg/cmd/create_job.go [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+// the current context on a user's KUBECONFIG.
+type CreateJobOptions struct {
+	configFlags       *genericclioptions.ConfigFlags
+	previewPrintFlags *genericclioptions.JSONYamlPrintFlags
+	printFlags        *genericclioptions.PrintFlags
+	genericclioptions.IOStreams
+
+	printObj func(obj runtime.Object) error
+
+	name string
```
Comment (mpon) [inline pkg/cmd/create_job.go:76]: @d-kuro How do you think about exporting or not the fields of the struct?

I found that it's not exported in sample plugin.
https://github.com/kubernetes/sample-cli-plugin/blob/master/pkg/cmd/ns.go#L49

But it does in kubectl.
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/get/get.go#L55-L85

File: pkg/cmd/create_job.go [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+
+	var printer printers.ResourcePrinter
+	if o.Preview {
+		printer, err = o.previewPrintFlags.ToPrinter(o.PreviewFormat)
+		if err != nil {
+			return err
+		}
+	}
+
+	info, err := fuzzyfinder.Infos(infos, printer, false, o.RawPreview)
```
Comment (mpon) [inline pkg/cmd/create_job.go:182]: I thought with `resource.Info` I could use the existing fuzzyfinder implementation.

File: pkg/cmd/create_job.go (hunk 8 above)
Comment (mpon) [reply pkg/cmd/create_job.go:198]: If this line is removed, it shows error, so I keep it.

```go
pkg/cmd/create_job.go:199:2: only one cuddle assignment allowed before if statement (wsl)
        if err != nil {
        ^
make: *** [lint] Error 1
```

File: pkg/cmd/create_job.go (hunk 12 above)
Comment (d-kuro) [reply pkg/cmd/create_job.go:76]: I don't think it's necessary to export unless it's referenced from an outside package.
The existing code I wrote is also an unnecessary export. I hope to fix it soon.

File: pkg/cmd/create_job.go (hunk 13 above)
Comment (d-kuro) [reply pkg/cmd/create_job.go:182]: cool! 👍 
I will fix the existing code.

File: pkg/cmd/create_job.go (hunk 8 above)
Comment (d-kuro) [reply pkg/cmd/create_job.go:198]: I see,  Thanks report!

**Pull request merged**